### PR TITLE
HPCC-29812 Mitigate rename failures on k8s with nfs backed storage.

### PR DESCRIPTION
**Pull request by jakesmith**
Occassionally, renaming a physical file that is on a mount backed by Azure Blob storage over NFS fails. That is ::rename returns -1 indicating an error. However, on inspection immediately after the failure, the rename has actually occurred, i.e. the src file no longer exists, and the dst does.

Add a mechanism and configuration option to allow CFile::rename to continue after failure if it can establish that the source filename has actually been renamed to the target filename.

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: system/jlib/jfile.cpp [hunk 1]
```diff
@@ -749,74 +749,94 @@ bool CFile::remove()
...
+            {
+                if (0 == dstStatErr) // dst now exists
+                {
+                    DBGLOG("%s-doRename(%s, %s) FAILED BUT DID RENAME (dst modtime=%s)", callerName, src, dst, dstModTimeStr.str());
+                    return;
+                }
+            }
         }
+        VStringBuffer errMsg("%s-doRename(%s, %s) FAILED [err=%d] ", callerName, src, dst, err);
+        errMsg.appendf("src stat:[err=%d", srcStatErr);
```
Comment (mckellyln) [inline system/jlib/jfile.cpp:831]: minor comment -
if srcStatErr is not 0, errno is almost surely ENOENT, but might still be interesting to log the errno.
and same for dstStatErr.

**Review by mckellyln — APPROVED** [review]
Looks fine to merge.
Just one minor comment.

File: system/jlib/jfile.cpp [hunk 2]
```diff
@@ -749,74 +749,94 @@ bool CFile::remove()
...
-    unsigned retries;
+    if ((unsigned)-1 != renameRetries)
+        return;
     // NB: potentially could be >1 thread here, but that's ok.
     try
     {
-        retries = getGlobalConfigSP()->getPropInt("expert/@numRenameRetries");
+        Owned<IPropertyTree> globalConfig = getGlobalConfigSP();
+        Owned<IPropertyTree> config = getComponentConfigSP();
+        manualRenameChk = globalConfig->getPropBool("expert/@manualRenameChk", config->getPropBool("expert/@manualRenameChk"));
```
Comment (ghalliday) [inline system/jlib/jfile.cpp:769]: The calls are in the wrong order.  This means the global config will take precedence over the component config.

**Review by ghalliday — COMMENTED** [review]
@jakesmith one minor bug.
Would there be a big disadvantage to enabling this by default (with 0 retries)?  What would be the downside?

File: system/jlib/jfile.cpp (hunk 2 above)
Comment (jakesmith) [reply system/jlib/jfile.cpp:769]: yes, agreed, will flip.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**